File: js/euler93.js
function euler93(){
    let maxStreak = 0;
    let res = "";
    let opsList = generateOps();
    let combs = generateCombs([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 4);
    let groups = [
        "xoxoxox", "(xox)oxox", "xo(xox)ox", "xoxo(xox)", "(xox)o(xox)",
        "(xoxox)ox", "((xox)ox)ox", "(xo(xox))ox",
        "xo(xoxox)", "xo((xox)ox)", "xo(xo(xox))"
    ];
    for (let comb of combs){
        let perms = generatePerms(comb);
        let results = getResults(perms, opsList, groups);
        let streak = getStreak(results);
        if (streak > maxStreak){
            maxStreak = streak;
            res = comb.join('');
        }
    }
    return res;
}

function generateOps(){
    const ops = ['+', '-', '/', '*'];
    const res = [];
    for (let x of ops){
        for (let y of ops){
            for (let z of ops){
                res.push([x, y, z]);
            }
        }
    }
    return res;
}

function generateCombs(nums, r){
    const combs = [];
    function backtrack(start, curr){
        if (curr.length == r){
            combs.push([...curr]);
            return;
        }
        for (let i = start; i < nums.length; i++){
            curr.push(nums[i]);
            backtrack(i + 1, curr);
            curr.pop();
        }
    }
    backtrack(0, []);
    return combs;
}

function generatePerms(nums){
    const perms = [];
    var seen = new Array(nums.length).fill(false);
    function backtrack(perm){
        if (perm.length == nums.length){
            perms.push([...perm]);
            return;
        }
        for (let i = 0; i < nums.length; i++){
            if (!seen[i]){
                perm.push(nums[i]);
                seen[i] = true;
                backtrack(perm);
                perm.pop();
                seen[i] = false;
            }
        }
    }
    backtrack([]);
    return perms;
}

function getResults(perms, ops, groups){
    const results = new Set();
    for (let perm of perms){
        for (let group of groups){
            for (let op of ops){
                let i = 0, j = 0;
                let resStr = "";
                for (let c of group){
                    if (c == "x") resStr += perm[i++];
                    else if (c == "o") resStr += op[j++];
                    else resStr += c;
                }
                let res = eval(resStr);
                if (isFinite(res) && res > 0 && res == Math.floor(res)) results.add(res);
            }
        }
    }
    return results;
}

function getStreak(vals){
    let streak = 0;
    while (vals.has(streak + 1)) streak++;
    return streak;
}

console.log(euler93());

File: go/euler93.go
package main
import "fmt"

func generateOps() [][]string {
    syms := []string{"+", "-", "*", "/"}
    ops := [][]string{}
    for i := 0; i < 4; i++ {
        for j := 0; j < 4; j++ {
            for k := 0; k < 4; k++ {
                op := []string{syms[i], syms[j], syms[k]}
                ops = append(ops, op)
            }
        }
    }
    return ops
}

func generateCombs(n int, k int) [][]int {
    combs := [][]int{}
    curr := []int{}
    
    var backtrack func(int)
    backtrack = func(idx int) {
        if len(curr) == k {
            comb := make([]int, k)
            copy(comb, curr)
            combs = append(combs, comb)
            return
        }
        for i := idx; i <= n; i++ {
            curr = append(curr, i)
            backtrack(i + 1)
            curr = curr[:len(curr) - 1]
        }
    }
    backtrack(0)
    return combs
}

func generatePerms(nums []int) [][]int {
    perms := [][]int{}
    used := make([]bool, len(nums))
    curr := []int{}
    
    var backtrack func(int)
    backtrack = func(idx int) {
        if idx == len(nums) {
            perm := make([]int, len(nums))
            copy(perm, curr)
            perms = append(perms, perm)
            return
        }
        for i := 0; i < len(nums); i++ {
            if !used[i] {
                curr = append(curr, nums[i])
                used[i] = true
                backtrack(idx + 1)
                curr = curr[:len(curr) - 1]
                used[i] = false
            }
        }
    }
    backtrack(0)
    return perms
}

func getStreak(nums []int, ops []string) int {
    perms := generatePerms(nums)
    seen := make(map[int]bool)
    groups := []string{
        "xxxx...", "xxx.x..", "xxx..x.", "xx.xx..", "xx.x.x.",
    }
    
    for _, perm := range perms {
        for _, op := range ops {
            for _, group := range groups {
                seen[eval(perm, op, group)] = true
            }
        }
    }
    for i := 1; ; i++ {
        if _, found := seen[i]; !found {
            return i - 1
        }
    }
}

func eval(nums []int, ops string, group string) int {
    stack := []float{}
    // TODO: push nums and apply postfix ops according to group structure
    for _, c := range group {
        // if c == x
        
        // if c == .
    }
    
    res := stack[0]
    if res < 0 {
        return -1
    }
    return res
}

func main() {
    res, maxStreak := 0, 0
    ops := generateOps()
    
    combs := generateCombs(9, 4)
    for _, comb := range combs {
        for _, op := range ops {
            streak := getStreak(comb, op)
            if streak > maxStreak {
                maxStreak = streak
                res = comb[0] * 1000 + comb[1] * 100 + comb[2] * 10 + comb[3]
            }
        }
    }
    fmt.Println(res)
}